pkg/streamingpromql/operators/function_over_instant_vector_test.go

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+	"testing"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/pooling"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// Most of the functionality of functions is tested through the test scripts in
+// pkg/streamingpromql/testdata.
+
+func TestFunctionOverInstantVector(t *testing.T) {
```

Comment:
What is this test trying to test? I'm not sure this is necessary - we can rely on the tests in the test scripts for the functions themselves.

Reply:
It doesn't test anything extra that wouldn't be covered by the functional tests, but it is a unit test to confirm it does what we expect. It tests two things:

- That both the `MetadataFunc` and `SeriesDataFunc` are called
- That `SeriesDataFunc` is called once per series

Reply:
I still don't think this is necessary - we can confirm that the functions are called by observing their effects (eg. that the label name has been dropped or that the values returned are equal to `acos(<value>)`).

Reply:
Whilst very unlikely, it is possible that we confirm the effects correctly but do something like call `SeriesDataFunc` more than once.

We could make sure that is unlikely by writing functional tests (ie the test scripts) for functions that are non-idempotent, but then we run the risk of not maintaining that test correctly (eg we refactor or remove it in the future).

I agree it's very unlikely (maybe even impossible), and that this test (and the others) don't necessarily add much, but I also don't see it as a huge burden to keep the extra testing and assurances.

Worst case in regards to these (likely redundant) tests, is that we remove them later when we need to change the way something works internally and still satisfies the existing functional test.